bin/mcb

```diff
@@ -63,13 +64,17 @@ end
 
 def show_all_commands(cmd, prefix)
   cmd.commands.each do |c|
-    aligned_cmd_name = (prefix + c.name).ljust(37, ' ')
-    coloured_name = Rainbow(aligned_cmd_name).green
-    puts "#{coloured_name} #{c.summary}"
+    show_this_command c, prefix
     show_all_commands(c, prefix + c.name + " ")
   end
 end
 
+def show_this_command(c, prefix)
+  aligned_cmd_name = (prefix + c.name).ljust(37, ' ')
+  coloured_name = Rainbow(aligned_cmd_name).green
+  puts "#{coloured_name} #{c.summary}"
+end
+
```

Comment:
How about this, which would replace both of these functions?

```
def show_command_help(cmd, prefix = '   ')
  full_cmd_name = "#{prefix} #{cmd.name}"

  if cmd != $mcb
    coloured_cmd_name = Rainbow(full_cmd_name).green
    puts '%<cmd>-46s %<summary>s' % { cmd: coloured_cmd_name, summary: cmd.summary }
  end

  cmd.commands.each { |c| show_command_help(c, full_cmd_name) }
end
```

Reply:
it doesn't quite behave as I'd wanted

[ ] it repeats inner command help, e.g. `bin/mcb az -h` repeats the description of az itself
[ ] it's lost the spacing between groups of commands
[ ] it includes an `mcb` prefix for all commands which is noisy and doesn't match the default help

Reply:
unified diff for easier patching:

```
diff --git a/bin/mcb b/bin/mcb
index abc575f..e947b90 100755
--- a/bin/mcb
+++ b/bin/mcb
@@ -47,12 +47,7 @@ $mcb = Cri::Command.define do
     puts cmd.help
     puts
     puts Rainbow("SUBCOMMANDS").red.bright
-    prefix = "    "
-    cmd.commands.each do |c|
-      show_this_command c, prefix unless cmd.name == 'mcb'
-      show_all_commands(c, prefix + c.name + " ")
-      puts
-    end
+    show_command_help(cmd)
     exit 0
   end
 
@@ -62,17 +57,15 @@ $mcb = Cri::Command.define do
   end
 end
 
-def show_all_commands(cmd, prefix)
-  cmd.commands.each do |c|
-    show_this_command c, prefix
-    show_all_commands(c, prefix + c.name + " ")
+def show_command_help(cmd, prefix = '   ')
+  full_cmd_name = "#{prefix} #{cmd.name}"
+
+  if cmd != $mcb
+    coloured_cmd_name = Rainbow(full_cmd_name).green
+    puts '%<cmd>-46s %<summary>s' % { cmd: coloured_cmd_name, summary: cmd.summary }
   end
-end
 
-def show_this_command(c, prefix)
-  aligned_cmd_name = (prefix + c.name).ljust(37, ' ')
-  coloured_name = Rainbow(aligned_cmd_name).green
-  puts "#{coloured_name} #{c.summary}"
+  cmd.commands.each { |c| show_command_help(c, full_cmd_name) }
 end
 
 MCB.load_commands($mcb, "#{lib_dir}/mcb/commands")
```

Reply:
Sorry, I figured those issues would be easy to fix. :) Here's a new patch:

```patch
diff --git a/bin/mcb b/bin/mcb
index abc575f0..bc964942 100755
--- a/bin/mcb
+++ b/bin/mcb
@@ -47,12 +47,9 @@ $mcb = Cri::Command.define do
     puts cmd.help
     puts
     puts Rainbow("SUBCOMMANDS").red.bright
-    prefix = "    "
-    cmd.commands.each do |c|
-      show_this_command c, prefix unless cmd.name == 'mcb'
-      show_all_commands(c, prefix + c.name + " ")
-      puts
-    end
+
+    show_command_help(cmd, top: true, group: cmd.name == 'mcb')
+
     exit 0
   end

@@ -62,17 +59,18 @@ $mcb = Cri::Command.define do
   end
 end

-def show_all_commands(cmd, prefix)
-  cmd.commands.each do |c|
-    show_this_command c, prefix
-    show_all_commands(c, prefix + c.name + " ")
+def show_command_help(cmd, prefix: '   ', top: false, group: false)
+  unless top
+    # We don't display top-level commands.
+    full_cmd_name = "#{prefix} #{cmd.name}"
+    coloured_cmd_name = Rainbow(full_cmd_name).green
+    puts '%<cmd>-46s %<summary>s' % { cmd: coloured_cmd_name, summary: cmd.summary }
   end
-end

-def show_this_command(c, prefix)
-  aligned_cmd_name = (prefix + c.name).ljust(37, ' ')
-  coloured_name = Rainbow(aligned_cmd_name).green
-  puts "#{coloured_name} #{c.summary}"
+  cmd.commands.sort_by(&:name).each do |c|
+    show_command_help(c, prefix: full_cmd_name)
+    puts "\n" if group
+  end
 end

 MCB.load_commands($mcb, "#{lib_dir}/mcb/commands")
```